_drafts/2024-07-25-testing-raise.md

```diff
@@ -0,0 +1,506 @@
+
+We have used JUnit 5 until now. However, we can switch to Kotest. **Kotest is a robust testing framework for Kotlin**, which is very close to ScalaTest for the Scala language. Kotest also has a set of tailored assertions for some of the available types in the Arrow library (see the [documentation](https://kotest.io/docs/assertions/arrow.html) for further details).
+
+So, let's translate the above tests in Kotest notation.
+
+```kotlin
+internal class CreatePortfolioUseCaseKotestTest : ShouldSpec({
+
+    val underTest = createPortfolioUseCase()
+
+    context("The create portfolio use case") {
+        should("create a portfolio for a user") {
+            val actualResult: Either<DomainError, PortfolioId> =
+                either {
+                    underTest.createPortfolio(CreatePortfolio(UserId("bob"), Money(1000.0)))
+                }
+
+            actualResult.shouldBeRight(PortfolioId("1"))
+        }
+    }
+})
+```
+
+This article is not a tutorial on how to use Kotest. However, we can notice that tests are not methods of the class, as in JUnit 5. Tests are lambdas with receiver given to the `ShouldSpec` class, one of the available contexts. If you extend the `ShouldSpec` class, you can access the `context` and `should` methods, declared functions in the `ShouldSpecRootScope`. The lambda you give to the `ShouldSpec` constructor is defined with the `ShouldSpecRootScope` context. As you might notice, we didn't use the `runTest` or `runBlocking` functions since the `should` method accepts a suspending function itself.
+
+In this test, we used the approach of converting the `Raise<E>.() -> A` context into an `() -> Either<E, A>`, and then we took advantage of the `shouldBeRight` assertions given by the `kotest-assertions-arrow` library.
+
+Unfortunately, Kotest has no extension to test a `Raise<E>.() -> A` function directly. So, if we don't want to convert it into an `Either<E, A>`, we need to use the approach with the `fold` method:
+
+```kotlin
+should("create a portfolio for a user (using fold)") {
+    fold(
+        block = { underTest.createPortfolio(CreatePortfolio(UserId("bob"), Money(1000.0))) },
+        recover = { fail("The use case should not fail") },
+        transform = { it.shouldBe(PortfolioId("1")) },
+    )
+}
+```
+
+So, we briefly introduce how to test a function that can raise an error of type `E`. What if the function calls another function that can raise an error, too? We should not depend on code outside the unit we're testing. At the end of the day, we focus on unit testing. Let's see what the available options are.
+
+## 3. Mocking the Raise DSL
+
+When dealing with unit tests and depending on a component outside our unit under test, we have at least two choices: fake objects and mocks. But before going deep into the topic, let's prepare the playground.
```

Comment:
Fake objects - you mean stubs?
I see you introduce stubs as defined by Martin Fowler, but it would make sense to make the connection between "fake objects" and stubs.

Reply:
Nope. I meant _fake objects_. It's a popular term, and I gave the definition made by Martin Fowler some lines later. When you implement the interface of an external dependency in the tests, it's called a "fake object"